File: backend-dotnet/src/Veha.Domain/Rules/StageRules.cs
using Veha.Domain.Common;
using Veha.Domain.Enums;

namespace Veha.Domain.Rules;

/// <summary>Правила переходов по стадиям (порт stage_rules.py). Только вперёд на
/// одну стадию или откат на одну назад с указанием причины.</summary>
public static class StageRules
{
    public static readonly Stage[] Order =
    [
        Stage.Presale,
        Stage.Survey,
        Stage.Design,
        Stage.Implementation,
        Stage.Pilot,
        Stage.Support,
        Stage.Closed,
    ];

    public static int StageIndex(Stage stage) => Array.IndexOf(Order, stage);

    public static Stage? NextStage(Stage stage)
    {
        var i = StageIndex(stage);
        return i + 1 < Order.Length ? Order[i + 1] : null;
    }

    public static Stage? PrevStage(Stage stage)
    {
        var i = StageIndex(stage);
        return i > 0 ? Order[i - 1] : null;
    }

    public static bool IsForward(Stage from, Stage to) => StageIndex(to) - StageIndex(from) == 1;

    public static bool IsBackward(Stage from, Stage to) => StageIndex(from) - StageIndex(to) == 1;

    /// <summary>Бросает DomainValidationException, если переход недопустим.</summary>
    public static void ValidateTransition(Stage from, Stage to, string? reason)
    {
        if (from == to)
            throw new DomainValidationException("Стадия не изменилась");
        if (IsForward(from, to))
            return;
        if (IsBackward(from, to))
        {
            if (string.IsNullOrWhiteSpace(reason))
                throw new DomainValidationException("Откат на предыдущую стадию требует указания причины");
            return;
        }
        throw new DomainValidationException(
            "Недопустимый переход: разрешено только вперёд на одну стадию " +
            "или откат на одну назад с указанием причины");
    }
}

File: backend/app/services/stage_rules.py
"""Чистые правила переходов по стадиям проекта (без БД — легко тестируется).

Переходы только последовательные вперёд на одну стадию и откат на одну назад
(с указанием причины). См. CLAUDE.md §4 «Проекты»."""

from __future__ import annotations

from app.models.enums import STAGE_ORDER, Stage


def stage_index(stage: Stage) -> int:
    return STAGE_ORDER.index(stage)


def next_stage(stage: Stage) -> Stage | None:
    i = stage_index(stage)
    return STAGE_ORDER[i + 1] if i + 1 < len(STAGE_ORDER) else None


def prev_stage(stage: Stage) -> Stage | None:
    i = stage_index(stage)
    return STAGE_ORDER[i - 1] if i > 0 else None


def is_forward(from_stage: Stage, to_stage: Stage) -> bool:
    return stage_index(to_stage) - stage_index(from_stage) == 1


def is_backward(from_stage: Stage, to_stage: Stage) -> bool:
    return stage_index(from_stage) - stage_index(to_stage) == 1


def validate_transition(from_stage: Stage, to_stage: Stage, *, reason: str | None) -> None:
    """Бросает ValueError с человекочитаемым сообщением, если переход недопустим."""
    if from_stage == to_stage:
        raise ValueError("Стадия не изменилась")
    if is_forward(from_stage, to_stage):
        return
    if is_backward(from_stage, to_stage):
        if not reason or not reason.strip():
            raise ValueError("Откат на предыдущую стадию требует указания причины")
        return
    raise ValueError(
        "Недопустимый переход: разрешено только вперёд на одну стадию "
        "или откат на одну назад с указанием причины"
    )
